### PR TITLE
 plan, executor: refine projection elimination when projection is the inner child of an outer join

### LINKED ISSUE #4997: [Randgen] Panic on optimizer_subquery_portable.yy
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
- download randgen https://launchpad.net/randgen
- build TIDB 1.0.1 (1.1.0 doesn't build sucessfully)
- run desired test: ```./gentest.pl --dsn dbi:mysql:host=127.0.0.1:port=4000:user=root:database=test --grammar=conf/optimizer_subquery_portable.yy --gendata --thread=1```

2. What did you expect to see?
Test passing

3. What did you see instead?
TIDB panic with :

```
panic: runtime error: index out of range

goroutine 8713 [running]:
github.com/pingcap/tidb/expression.(*Column).Eval(0xc420686ea0, 0xc4215d4160, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/pingcap/tidb/expression/column.go:197 +0x99
github.com/pingcap/tidb/expression.EvalBool(0xc420ffc250, 0x1, 0x1, 0xc4215d4160, 0x3, 0x3, 0x18ac880, 0xc421ef2780, 0x0, 0xc4222ba000, ...)
        /go/src/github.com/pingcap/tidb/expression/expression.go:103 +0xac
github.com/pingcap/tidb/executor.(*HashJoinExec).constructMatchedRows(0xc42172a900, 0xc42236eb40, 0xc422762a80, 0x2, 0x2, 0x2, 0x18ac880, 0xc421ef2780, 0x1, 0x0)
        /go/src/github.com/pingcap/tidb/executor/join.go:388 +0x3fb
github.com/pingcap/tidb/executor.(*HashJoinExec).joinOneBigRow(0xc42172a900, 0xc42236eb40, 0xc422762a80, 0x2, 0x2, 0xc420eb9530, 0xc421b8d140)
        /go/src/github.com/pingcap/tidb/executor/join.go:347 +0x3e1
github.com/pingcap/tidb/executor.(*HashJoinExec).runJoinWorker(0xc42172a900, 0x3)
        /go/src/github.com/pingcap/tidb/executor/join.go:316 +0x2a9
created by github.com/pingcap/tidb/executor.(*HashJoinExec).prepare
        /go/src/github.com/pingcap/tidb/executor/join.go:250 +0x52a
```

4. What version of TiDB are you using (`tidb-server -V`)?
1.0.1

### PR DESCRIPTION
**Pull request by XuHuaiyu**
When a `Projection upon a Join` be eliminated, if the schema of Projection 
is not consistent with the schema of Join, the `DefaultValues` of Join
should be reset.
fix #4997 #4996 
PTAL @coocood @zz-jason @hanfei1991

### CODE REVIEW

File: plan/physical_plans.go [hunk 1]
```diff
@@ -561,6 +596,7 @@ func buildJoinSchema(joinType JoinType, join Plan, outerID int) *expression.Sche
 		newSchema.Append(join.Schema().Columns[join.Schema().Len()-1])
 		return newSchema
 	}
+	resetDefaultValues(joinType, join)
```
Comment (zz-jason) [inline plan/physical_plans.go]: how about `resetDefaultValues(joinType, join, outerID)`, then in function `resetDefaultValues` we can get `childSchemaLen` directly by `join.Children()[outerID].Schema().Len()`

Comment (XuHuaiyu) [reply plan/physical_plans.go]: maybe no difference, the function still has to check LeftOuterJoin and RightOuterJoin @zz-jason

File: plan/eliminate_projection.go [hunk 2]
```diff
@@ -157,14 +158,49 @@ func (pe *projectionEliminater) eliminate(p LogicalPlan, replace map[string]*exp
...
+	case LeftOuterJoin:
+		joinInnerChild = join.Children()[1]
+	case RightOuterJoin:
+		joinInnerChild = join.Children()[0]
+	}
+	if joinInnerChild != prj {
+		return join.DefaultValues
+	}
+	var schemaIdxMap map[int]int
+	joinInnerChildSchema := joinInnerChild.Schema().Columns
```
Comment (coocood) [inline plan/eliminate_projection.go]: I think `prj` is easier to read.

**Comment by coocood** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

**Pull request merged**